Fix mergeable section flags.

jacobly0 opened the pull request:
Set the correct flags and entsize on the mergeable ELF sections to avoid an assembler error.  In particular, the `cst8` &rightarrow; `cst16` changes seem particularly important to avoid the 16 byte constants from being split and merged incorrectly.  If merging is not actually wanted, then they should be placed in the normal `.rodata` section instead.

jacobly0 committed:
Fix mergeable section flags.

nojb commented:
Could you give some more information about what kind of error you are trying to address here (architecture, operating system, versions involved, reproduction case, etc)? Thanks!

jacobly0 commented:
I'm testing on x86_64 linux.  After looking into it a bit more, the error only occurs when using the assembler integrated into clang 11.0.0, which was just released.

```
$ cat test.s
	.section .rodata.cst8, "a", @progbits

$ gcc-10.2.0 --version; gcc-10.2.0 -c test.s
gcc-10.2.0 (Gentoo 10.2.0-r2 p3) 10.2.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ clang-10 --version; clang-10 -c test.s
clang version 10.0.1 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm/10/bin

$ clang-11 --version; clang-11 -c test.s
clang version 11.0.0
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm/11/bin
test.s:1:2: error: changed section flags for .rodata.cst8, expected: 0x12
        .section .rodata.cst8, "a", @progbits
        ^
test.s:1:2: error: changed section entsize for .rodata.cst8, expected: 8
        .section .rodata.cst8, "a", @progbits
        ^
```

hannesm commented:
a similar patch has been applied to the FreeBSD package of OCaml (still at 4.05.1 due to safe-string) -- take a look at https://github.com/freebsd/freebsd-ports/commit/0076f28d1fbe3d1a3f3470f66a6b3416433eb0bc#diff-736aa35d0fe6c87e7ad0e4a0f1228d65eddaa1b6a6007e5d04c9742857039c88 if interested.

hannesm commented:
In the FreeBSD bugzilla a bug to upgrade OCaml to 4.11.1 is open, which uses two patches for the issue reported here:
- [bug 250408](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=250408)
- [patch 1](https://bugs.freebsd.org/bugzilla/attachment.cgi?id=218814&action=diff)
- [patch 2](https://bugs.freebsd.org/bugzilla/attachment.cgi?id=218815&action=diff)

nojb reviewed (commented):
Overall the patch looks reasonable.

The change `cst8 => cst16` seems pertinent independently of this bug report since we are emitting 16-byte constants in those sections.

I am running a [precheck](https://ci.inria.fr/ocaml/job/precheck/523/) to make sure this doesn't cause regressions in the existing ports.

Something I don't understand is why the mergeability flag `M` is mandated by the `clang` assembler (even if it may be a good idea to use).

jacobly0 commented:
By convention, `.rodata.cst8` is the name of a mergeable section, so clang is expecting you to set the mergeable flag so that it matches other instances of `.rodata.cst8`.  If you don't set the mergeable flag, then the size of the constant isn't relevant and you might as well just put it in `.rodata` instead, which can't have the mergeable flag.

nojb commented:
> By convention, `.rodata.cst8` is the name of a mergeable section, so clang is expecting you to set the mergeable flag so that it matches other instances of `.rodata.cst8`. If you don't set the mergeable flag, then the size of the constant isn't relevant and you might as well just put it in `.rodata` instead, which can't have the mergeable flag.

Thanks for the explanation. Is there any downside to using a plain `.rodata` section apart from the lack of mergeability ?

jacobly0 commented:
The only reason for the separate section is for merging, so yes that is the only difference.  Although one minor detail is that keeping constants of the same size together happens to reduce space wasted by alignment.

nojb reviewed (approved):
I am approving because I am satisfied that the patch is correct (also: `precheck` passed).

Let us wait a couple of days before merging to give a chance to other developers to chime in.

@jacobly0 please add a `Changes` entry as it will be useful for anyone running into this `clang` issue.

Thanks!

nojb commented on runtime/amd64.S:
```diff
@@ -679,7 +679,7 @@ G(caml_system__frametable):
+        .section    .rodata.cst16,"aM",@progbits,16
 #endif
         .globl  G(caml_negf_mask)
         .align  SIXTEEN_ALIGN
```
@jacobly0 Is it the case that this alignment directive (and the one below) are no longer needed?

jacobly0 replied:
At least one `.align` directive is needed per section, as the largest alignment that appears in a section defines the alignment of the entire section, and the default is 1 byte alignment.  The fourth argument to a `.section` directive (with `"M"` flag but not `"S"` flag) defines how big each entry in the section is, and is independent of alignment (As a random example, ELF64 `.rela.*` sections have an alignment of 8 bytes but an entsize of 24 bytes).  This means that only the second `.align` directive is superfluous here, since there happens to be exactly 16 bytes between it and the previous one.  Since you brought up alignment, I also added a missing one in the other changed file.

xavierleroy commented on runtime/amd64.S:
```diff
@@ -679,14 +679,13 @@ G(caml_system__frametable):
 G(caml_negf_mask):
         .quad   0x8000000000000000, 0
         .globl  G(caml_absf_mask)
-        .align  SIXTEEN_ALIGN
```
I like to be very explicit about alignment.  Please don't remove this `.align` even if it happens to be redundant.

xavierleroy reviewed (commented):
Well spotted and nicely fixed, thanks.  There is one unrelated change that I find unwarranted and would like you to revert, even though it was suggested by a reviewer.

xavierleroy reviewed (approved):
Excellent, thank you.

gasche commented:
Note: I'm thinking of backporting this patch in some earlier release branches, to make them buildable using newer C toolchains.

hannesm commented:
> Note: I'm thinking of backporting this patch in some earlier release branches, to make them buildable using newer C toolchains.

as a heads up to release management team (//cc @Octachron): I think this is indeed sensible to backport to 4.10 & 4.11 branches to enable them to be built with clang 11. Thanks.

bschommer commented:
I can confirm that the fix is enough to get 4.11 working with a newer clang version.

Octachron commented:
The fix does sounds completely fine for 4.11.2 .
Moreover, @hannesm mentioned that this patch is needed for compiling on FreeBSD/amd64 which is a tier 1 platform.
I am therefore thinking of integrating it to 4.10.2 as the "Tier 1 platform support fix" release .

hannesm commented:
FWIW, this has been included in 4.10.2 (i.e. merged into the 4.10 branch), but is not yet merged into the 4.11 branch -- which would be appreciated now that a 4.11 release is on the horizon... //cc @Octachron

Octachron commented:
Indeed, I will cherry-pick this commit to the 4.11 branch as the only non-MacOS/arm64 bugfix missing when compared to 4.10.2.
